netstat/netstat.1
.\" Copyright (c) 1983 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted
.\" provided that the above copyright notice and this paragraph are
.\" duplicated in all such forms and that any documentation,
.\" advertising materials, and other materials related to such
.\" distribution and use acknowledge that the software was developed
.\" by the University of California, Berkeley.  The name of the
.\" University may not be used to endorse or promote products derived
.\" from this software without specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
.\" WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)netstat.1	6.8 (Berkeley) 9/20/88
.\"
.TH NETSTAT 1 "September 20, 1988"
.UC 5
.SH NAME
netstat \- show network status
.SH SYNOPSIS
.B netstat
[ 
.B \-Aan
] [
.B \-f
.I address_family
] [
.I system
] [
.I core
]
.br
.B netstat
[ 
.B \-himnrs
] [
.B \-f
.I address_family
] [
.I system
] [
.I core
]
.br
.B netstat
[ 
.B \-n
] [
.B \-I
.I interface
]
.I interval
[
.I system
] [
.I core
]
.br
.B netstat
[ 
.B \-p
.I protocol
] [
.I system
] [
.I core
]
.SH DESCRIPTION
The
.I netstat 
command symbolically displays the contents of various network-related
data structures.
There are a number of output formats,
depending on the options for the information presented.
The first form of the command displays a list of active sockets for
each protocol.
The second form presents the contents of one of the other network
data structures according to the option selected.
Using the third form, with an 
.I interval
specified,
.I netstat
will continuously display the information regarding packet
traffic on the configured network interfaces.
The fourth form displays statistics about the named protocol.
.PP
The options have the following meaning:
.TP 
.B \-A
With the default display,
show the address of any protocol control blocks associated with sockets; used
for debugging.
.TP
.B \-a
With the default display,
show the state of all sockets; normally sockets used by
server processes are not shown.
.B \-d
With either interface display (option
.B \-i
or an interval, as described below),
show the number of dropped packets.
.TP
.B \-h
Show the state of the IMP host table.
.TP
.B \-i
Show the state of interfaces which have been auto-configured
(interfaces statically configured into a system, but not
located at boot time are not shown).
.TP
.BI \-I " interface"
Show information only about this interface;
used with an
.I interval
as described below.
.TP
.B \-m
Show statistics recorded by the memory management routines
(the network manages a private pool of memory buffers).
.TP
.B \-n
Show network addresses as numbers (normally 
.I netstat
interprets addresses and attempts to display them
symbolically).
This option may be used with any of the display formats.
.TP
.BI \-p " protocol"
Show statistics about 
.IR protocol ,
which is either a well-known name for a protocol or an alias for it.  Some
protocol names and aliases are listed in the file 
.IR /etc/protocols .
A null response typically means that there are no interesting numbers to 
report.
The program will complain if
.I protocol
is unknown or if there is no statistics routine for it.
.TP
.B \-s
Show per-protocol statistics.
.TP
.B \-r
Show the routing tables.
When
.B \-s
is also present, show routing statistics instead.
.TP
.BI \-f " address_family"
Limit statistics or address control block reports to those
of the specified
.IR address\ family .
The following address families
are recognized:
.IR inet ,
for
.BR AF_INET ,
.IR ns ,
for
.BR AF_NS ,
and
.IR unix ,
for
.BR AF_UNIX .
.PP
The arguments, 
.I system
and
.I core
allow substitutes for the defaults ``/vmunix'' and ``/dev/kmem''.
.PP
The default display, for active sockets, shows the local
and remote addresses, send and receive queue sizes (in bytes), protocol,
and the internal state of the protocol.
Address formats are of the form ``host.port'' or ``network.port''
if a socket's address specifies a network but no specific host address.
When known the host and network addresses are displayed symbolically
according to the data bases
.I /etc/hosts
and
.IR /etc/networks ,
respectively.  If a symbolic name for an address is unknown, or if
the 
.B \-n
option is specified, the address is printed numerically, according
to the address family.
For more information regarding 
the Internet ``dot format,''
refer to 
.IR inet (3N).
Unspecified,
or ``wildcard'', addresses and ports appear as ``*''.  
.PP
The interface display provides a table of cumulative
statistics regarding packets transferred, errors, and collisions.
The network addresses of the interface
and the maximum transmission unit (``mtu'') are also displayed.
.PP
The routing table display indicates the available routes and
their status.  Each route consists of a destination host or network
and a gateway to use in forwarding packets.  The flags field shows
the state of the route (``U'' if ``up''), whether the route
is to a gateway (``G''), whether the route was created dynamically
by a redirect (``D''), and whether the route has been modified
by a redirect (``M'').  Direct routes are created for each
interface attached to the local host;
the gateway field for such entries shows the address of the outgoing interface.
The refcnt field gives the
current number of active uses of the route.  Connection oriented
protocols normally hold on to a single route for the duration of
a connection while connectionless protocols obtain a route while sending
to the same destination.
The use field provides a count of the number of packets
sent using that route.  The interface entry indicates the network
interface utilized for the route.
.PP
When 
.I netstat
is invoked with an
.I interval
argument, it displays a running count of statistics related to
network interfaces.  This display consists of a
column for the primary interface
(the first interface found during autoconfiguration)
and a column summarizing
information for all interfaces.
The primary interface may be replaced with another interface with the
.I \-I
option.
The first line of each screen of information contains a summary since the
system was last rebooted.  Subsequent lines of output show values
accumulated over the preceding interval.
.SH SEE ALSO
iostat(1),
vmstat(1),
hosts(5),
networks(5),
protocols(5),
services(5),
trpt(8)
.SH BUGS
The notion of errors is ill-defined.  Collisions mean
something else for the IMP.


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


usr/src/usr.bin/systat/netstat.c
#ifndef lint
static char sccsid[] = "@(#)netstat.c	1.1 (Berkeley) %G%";
#endif

/*
 * netstat
 */
#include "systat.h"

#include <netdb.h>
#include <nlist.h>
#include <signal.h>
#include <arpa/inet.h>

#include <sys/types.h>
#include <sys/file.h>
#include <sys/socket.h>
#include <sys/socketvar.h>
#include <sys/mbuf.h>
#include <sys/protosw.h>

#include <net/route.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/in_pcb.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp_var.h>
#include <netinet/ip_var.h>
#include <netinet/tcp.h>
#include <netinet/tcpip.h>
#include <netinet/tcp_seq.h>
#define TCPSTATES
#include <netinet/tcp_fsm.h>
#include <netinet/tcp_timer.h>
#include <netinet/tcp_var.h>
#include <netinet/tcp_debug.h>
#include <netinet/udp.h>
#include <netinet/udp_var.h>

#define	streq(a,b)	(strcmp(a,b)==0)

WINDOW *
opennetstat()
{

	sethostent(1);
	setnetent(1);
	return (subwin(stdscr, LINES-5-1, 0, 5, 0));
}

struct netinfo {
	struct	netinfo *ni_forw, *ni_prev;
	short	ni_line;		/* line on screen */
	short	ni_seen;		/* 0 when not present in list */
	short	ni_flags;
#define	NIF_LACHG	0x1		/* local address changed */
#define	NIF_FACHG	0x2		/* foreign address changed */
	short	ni_state;		/* tcp state */
	char	*ni_proto;		/* protocol */
	struct	in_addr ni_laddr;	/* local address */
	long	ni_lport;		/* local port */
	struct	in_addr	ni_faddr;	/* foreign address */
	long	ni_fport;		/* foreign port */
	long	ni_rcvcc;		/* rcv buffer character count */
	long	ni_sndcc;		/* snd buffer character count */
};

static struct {
	struct	netinfo *ni_forw, *ni_prev;
} netcb;

static	int aflag = 0;
static	int nflag = 0;
static	int lastrow = 1;
static	char *inetname();

closenetstat(w)
        WINDOW *w;
{
	register struct netinfo *p, *q;

	endhostent();
	endnetent();
	p = netcb.ni_forw;
	while (p != (struct netinfo *)&netcb) {
		if (p->ni_line != -1)
			lastrow--;
		p->ni_line = -1;
		p = p->ni_forw;
	}
        if (w != NULL) {
		wclear(w);
		wrefresh(w);
		delwin(w);
	}
}

static struct nlist nlst[] = {
#define	X_TCB	0
	{ "_tcb" },
#define	X_UDB	1
	{ "_udb" },
        { "" },
};

static	long *ports;
static	int nports;
static	struct in_addr *hosts;
static	int nhosts;
static	int protos;
#define	TCP	0x1
#define	UDP	0x2

initnetstat()
{
        register  i;

	nlist("/vmunix", nlst);
	if (nlst[X_TCB].n_value == 0) {
		error("No symbols in namelist");
		return;
	}
	netcb.ni_forw = netcb.ni_prev = (struct netinfo *)&netcb;
	selectproto(0);
	selectport(-1);
	selecthost(0);
}

static
selectproto(proto)
	char *proto;
{
	int new = protos;

	if (proto == 0 || streq(proto, "all"))
		new = TCP|UDP;
	else if (streq(proto, "tcp"))
		new = TCP;
	else if (streq(proto, "udp"))
		new = UDP;
	return (new != protos, protos = new);
}

static
selectport(port)
	long port;
{
	register long *p;

	if (port == -1) {
		if (nports == 1)
			return (0);
		if (ports)
			free((char *)ports);
		ports = (long *)malloc(sizeof (long));
		*ports = -1;
		nports = 1;
		return (1);
	}
	for (p = ports; *p != -1; p++)
		if (*p == port)
			return (0);
	ports = (long *)realloc(ports, (nports + 1)*sizeof (long));
	ports[nports-1] = port;
	ports[nports++] = -1;
	return (1);
}

static
checkport(inp)
	register struct inpcb *inp;
{
	register long *p;

	for (p = ports; *p != -1; p++)
		if (*p == inp->inp_lport || *p == inp->inp_fport)
			return (1);
	return (0);
}

static
selecthost(in)
	struct in_addr *in;
{
	register struct in_addr *p;

	if (in == 0) {
		if (nhosts == 1)
			return (0);
		if (hosts)
			free((char *)hosts);
		hosts = (struct in_addr *)malloc(sizeof (struct in_addr));
		hosts->s_addr = -1;
		nhosts = 1;
		return (1);
	}
	for (p = hosts; p->s_addr != -1; p++)
		if (p->s_addr == in->s_addr)
			return (0);
	hosts = (struct in_addr *)realloc(hosts,
	    (nhosts + 1)*sizeof (struct in_addr));
	hosts[nhosts-1] = *in;
	hosts[nhosts++].s_addr = -1;
	return (1);
}

static
checkhost(inp)
	register struct inpcb *inp;
{
	register struct in_addr *p;

	for (p = hosts; p->s_addr != -1; p++)
		if (p->s_addr == inp->inp_laddr.s_addr ||
		    p->s_addr == inp->inp_faddr.s_addr)
			return (1);
	return (0);
}

fetchnetstat()
{
	register struct inpcb *prev, *next;
	register struct netinfo *p;
	struct inpcb inpcb;
	struct socket sockb;
	struct tcpcb tcpcb;
	off_t off;
	int istcp;

	if (nlst[X_TCB].n_value == 0)
		return;
	for (p = netcb.ni_forw; p != (struct netinfo *)&netcb; p = p->ni_forw)
		p->ni_seen = 0;
	if (protos&TCP)
		off = nlst[X_TCB].n_value, istcp = 1;
	else if (protos&UDP)
		off = nlst[X_UDB].n_value, istcp = 0;
again:
	lseek(kmem, off, L_SET);
	read(kmem, &inpcb, sizeof (struct inpcb));
	prev = (struct inpcb *)off;
	for (; inpcb.inp_next != (struct inpcb *)off; prev = next) {
		next = inpcb.inp_next;
		lseek(kmem, (off_t)next, L_SET);
		read(kmem, &inpcb, sizeof (inpcb));
		if (inpcb.inp_prev != prev) {
			p = netcb.ni_forw;
			for (; p != (struct netinfo *)&netcb; p = p->ni_forw)
				p->ni_seen = 1;
			error("Kernel state in transition");
			return;
		}
		if (!aflag && inet_lnaof(inpcb.inp_laddr) == INADDR_ANY)
			continue;
		if (nhosts != 1 && !checkhost(&inpcb))
			continue;
		if (nports != 1 && !checkport(&inpcb))
			continue;
		lseek(kmem, (off_t)inpcb.inp_socket, L_SET);
		read(kmem, &sockb, sizeof (sockb));
		lseek(kmem, (off_t)inpcb.inp_ppcb, L_SET);
		if (istcp) {
			read(kmem, &tcpcb, sizeof (tcpcb));
			enter(&inpcb, &sockb, tcpcb.t_state, "tcp");
		} else
			enter(&inpcb, &sockb, 0, "udp");
	}
	if (istcp && (protos&UDP)) {
		istcp = 0;
		off = nlst[X_UDB].n_value;
		goto again;
	}
}

static
enter(inp, so, state, proto)
	register struct inpcb *inp;
	register struct socket *so;
	int state;
	char *proto;
{
	register struct netinfo *p, *match = 0;
	int matchwild = 3, wildcard;

	/*
	 * Must mimic in_pcblookup to match partial connections.
	 */
	for (p = netcb.ni_forw; p != (struct netinfo *)&netcb; p = p->ni_forw) {
		if (!streq(proto, p->ni_proto))
			continue;
		if (p->ni_lport != inp->inp_lport)
			continue;
		wildcard = 0;
		if (p->ni_laddr.s_addr != INADDR_ANY) {
			if (inp->inp_laddr.s_addr == INADDR_ANY)
				wildcard++;
			else if (p->ni_laddr.s_addr != inp->inp_laddr.s_addr)
				continue;
		} else {
			if (inp->inp_laddr.s_addr != INADDR_ANY)
				wildcard++;
		}
		if (p->ni_faddr.s_addr != INADDR_ANY) {
			if (inp->inp_faddr.s_addr == INADDR_ANY)
				wildcard++;
			else if (p->ni_faddr.s_addr != inp->inp_faddr.s_addr ||
			    p->ni_fport != inp->inp_fport)
				continue;
		} else {
			if (inp->inp_faddr.s_addr != INADDR_ANY)
				wildcard++;
		}
		if (wildcard < matchwild) {
			match = p;
			matchwild = wildcard;
			if (matchwild == 0)
				break;
		}
	}
	if (match == 0) {
		match = (struct netinfo *)malloc(sizeof (*p));
		insque(match, &netcb);
		match->ni_line = -1;
		match->ni_flags = NIF_LACHG|NIF_FACHG;
	}
	if (match->ni_laddr.s_addr != inp->inp_laddr.s_addr ||
	    match->ni_lport != inp->inp_lport)
		match->ni_flags |= NIF_LACHG;
	match->ni_laddr = inp->inp_laddr;
	match->ni_lport = inp->inp_lport;
	if (match->ni_faddr.s_addr != inp->inp_faddr.s_addr ||
	    match->ni_fport != inp->inp_fport)
		match->ni_flags |= NIF_FACHG;
	match->ni_faddr = inp->inp_faddr;
	match->ni_fport = inp->inp_fport;
	match->ni_proto = proto;
	match->ni_rcvcc = so->so_rcv.sb_cc;
	match->ni_sndcc = so->so_snd.sb_cc;
	match->ni_state = state;
	match->ni_seen = 1;
}

/* column locations */
#define	LADDR	0
#define	FADDR	LADDR+23
#define	PROTO	FADDR+23
#define	RCVCC	PROTO+6
#define	SNDCC	RCVCC+7
#define	STATE	SNDCC+7

labelnetstat()
{

        if (nlst[X_TCB].n_type == 0)
                return;
        wmove(wnd, 0, 0); wclrtobot(wnd);
	mvwaddstr(wnd, 0, LADDR, "Local Address");
	mvwaddstr(wnd, 0, FADDR, "Foreign Address");
	mvwaddstr(wnd, 0, PROTO, "Proto");
	mvwaddstr(wnd, 0, RCVCC, "Recv-Q");
	mvwaddstr(wnd, 0, SNDCC, "Send-Q");
	mvwaddstr(wnd, 0, STATE, "(state)"); 
}

shownetstat()
{
	register struct netinfo *p, *q;

	/*
	 * First, delete any connections that have gone
	 * away and adjust the position of connections
	 * below to reflect the deleted line.
	 */
	p = netcb.ni_forw;
	while (p != (struct netinfo *)&netcb) {
		if (p->ni_line == -1 || p->ni_seen) {
			p = p->ni_forw;
			continue;
		}
		wmove(wnd, p->ni_line, 0); wdeleteln(wnd);
		q = netcb.ni_forw;
		for (; q != (struct netinfo *)&netcb; q = q->ni_forw)
			if (q != p && q->ni_line > p->ni_line)
				q->ni_line--;
		lastrow--;
		q = p->ni_forw;
		remque(p);
		free((char *)p);
		p = q;
	}
	wmove(wnd, lastrow, 0); wclrtobot(wnd);
	/*
	 * Update existing connections and add new ones.
	 */
	for (p = netcb.ni_forw; p != (struct netinfo *)&netcb; p = p->ni_forw) {
		if (p->ni_line == -1) {
			/*
			 * Add a new entry if possible.
			 */
			if (lastrow >= wnd->_maxy)
				continue;
			p->ni_line = lastrow++;
			p->ni_flags |= NIF_LACHG|NIF_FACHG;
		}
		if (p->ni_flags & NIF_LACHG) {
			wmove(wnd, p->ni_line, LADDR);
			inetprint(&p->ni_laddr, p->ni_lport, p->ni_proto);
			p->ni_flags &= ~NIF_LACHG;
		}
		if (p->ni_flags & NIF_FACHG) {
			wmove(wnd, p->ni_line, FADDR);
			inetprint(&p->ni_faddr, p->ni_fport, p->ni_proto);
			p->ni_flags &= ~NIF_FACHG;
		}
		mvwaddstr(wnd, p->ni_line, PROTO, p->ni_proto);
		mvwprintw(wnd, p->ni_line, RCVCC, "%6d", p->ni_rcvcc);
		mvwprintw(wnd, p->ni_line, SNDCC, "%6d", p->ni_sndcc);
		if (streq(p->ni_proto, "tcp"))
			if (p->ni_state < 0 || p->ni_state >= TCP_NSTATES)
				mvwprintw(wnd, p->ni_line, STATE, "%d",
				    p->ni_state);
			else
				mvwaddstr(wnd, p->ni_line, STATE,
				    tcpstates[p->ni_state]);
		wclrtoeol(wnd);
	}
}

/*
 * Pretty print an Internet address (net address + port).
 * If the nflag was specified, use numbers instead of names.
 */
static
inetprint(in, port, proto)
	register struct in_addr *in;
	int port;
	char *proto;
{
	struct servent *sp = 0;
	char line[80], *cp, *index();

	sprintf(line, "%.*s.", 16, inetname(*in));
	cp = index(line, '\0');
	if (!nflag && port)
		sp = getservbyport(port, proto);
	if (sp || port == 0)
		sprintf(cp, "%.8s", sp ? sp->s_name : "*");
	else
		sprintf(cp, "%d", ntohs((u_short)port));
	/* pad to full column to clear any garbage */
	cp = index(line, '\0');
	while (cp - line < 22)
		*cp++ = ' ';
	*cp = '\0';
	waddstr(wnd, line);
}

/*
 * Construct an Internet address representation.
 * If the nflag has been supplied, give 
 * numeric value, otherwise try for symbolic name.
 */
static char *
inetname(in)
	struct in_addr in;
{
	char *cp = 0;
	static char line[50];
	struct hostent *hp;
	struct netent *np;

	if (!nflag && in.s_addr != INADDR_ANY) {
		int net = inet_netof(in);
		int lna = inet_lnaof(in);

		if (lna == INADDR_ANY) {
			np = getnetbyaddr(net, AF_INET);
			if (np)
				cp = np->n_name;
		}
		if (cp == 0) {
			hp = gethostbyaddr(&in, sizeof (in), AF_INET);
			if (hp)
				cp = hp->h_name;
		}
	}
	if (in.s_addr == INADDR_ANY)
		strcpy(line, "*");
	else if (cp)
		strcpy(line, cp);
	else {
		in.s_addr = ntohl(in.s_addr);
#define C(x)	((x) & 0xff)
		sprintf(line, "%u.%u.%u.%u", C(in.s_addr >> 24),
			C(in.s_addr >> 16), C(in.s_addr >> 8), C(in.s_addr));
	}
	return (line);
}

cmdnetstat(cmd, args)
        char *cmd, *args;
{
	register struct netinfo *p;
	int omask;

	omask = sigblock(sigmask(SIGALRM));
	if (prefix(cmd, "all")) {
		aflag = !aflag;
		goto fixup;
	}
	if  (prefix(cmd, "numbers") || prefix(cmd, "names")) {
		int new;

		new = prefix(cmd, "numbers");
		if (new == nflag)
			goto done;
		p = netcb.ni_forw;
		for (; p != (struct netinfo *)&netcb; p = p->ni_forw) {
			if (p->ni_line == -1)
				continue;
			p->ni_flags |= NIF_LACHG|NIF_FACHG;
		}
		nflag = new;
		goto redisplay;
	}
	if (prefix(cmd, "tcp") || prefix(cmd, "udp")) {
		if (selectproto(cmd))
			goto fixup;
		goto done;
	}
	if (prefix(cmd, "protocol")) {
		if (selectproto(args))
			goto fixup;
		goto done;
	}
	if (prefix(cmd, "port")) {
		struct servent *sp;
		long port;

		sp = getservbyname(args,
			protos == TCP ? "tcp" : protos == UDP ? "udp" : 0);
		if (sp == 0) {
			port = atoi(args);
			if (port <= 0) {
				error("%s: unknown port", args);
				goto done;
			}
			port = htons((u_short)port);
		} else
			port = sp->s_port;
		if (selectport(port))
			goto fixup;
		goto done;
	}
	if (prefix(cmd, "host")) {
		struct hostent *hp;
		struct in_addr in;

		hp = gethostbyname(args);
		if (hp == 0) {
			in.s_addr = inet_addr(args);
			if (in.s_addr == -1) {
				error("%s: unknown host", args);
				goto done;
			}
		} else
			in = *(struct in_addr *)hp->h_addr;
		if (selecthost(&in))
			goto fixup;
		goto done;
	}
	if (prefix(cmd, "reset")) {
		if (selectproto(0) | selecthost(0) | selectport(-1))
			goto fixup;
		goto done;
	}
	sigsetmask(omask);
	return (0);
fixup:
	fetchnetstat();
redisplay:
	shownetstat();
	refresh();
done:
	sigsetmask(omask);
	return (1);
}
